File: time.go
package main

import (
	"time"

	"github.com/yuin/gopher-lua"
)

func RegisterTime(env *Environment) {
	env.RegisterNewType("time", map[string]lua.LGFunction{
		"now": func(L *lua.LState) int {
			env.Yield()
			L.Push(lua.LNumber(time.Now().UnixMilli()))
			return 1
		},
		"sleep": func(L *lua.LState) int {
			AsyncRun(env, func() struct{} {
				time.Sleep(time.Duration(float64(L.CheckNumber(1)) * float64(time.Millisecond)))
				return struct{}{}
			})
			return 0
		},
		"format": func(L *lua.LState) int {
			env.Yield()
			L.Push(lua.LString(time.UnixMilli(int64(L.CheckNumber(1))).Format(time.RFC3339)))
			return 1
		},
	}, map[string]lua.LValue{
		"millisecond": lua.LNumber(1),
		"second":      lua.LNumber(1000),
		"minute":      lua.LNumber(60 * 1000),
		"hour":        lua.LNumber(60 * 60 * 1000),
	})
}

File: internal/testdata/time.lua
a = time.now()
b = time.now()

assert.le(a, b)
assert.lt(b-a, 200*time.millisecond)

c = time.now()
time.sleep(100 * time.millisecond)
d = time.now()

assert.ge(d-c, 100*time.millisecond)
assert.lt(d-c, 300*time.millisecond)

assert.eq(os.getenv("TZ"), "UTC")
assert.eq(time.format(1136214245000), "2006-01-02T15:04:05+0000")
assert.eq(time.format(1136214245000, "%Y/%m/%d"), "2006/01/02")
assert.eq(time.format(1136214245000, "%H:%M:%S"), "15:04:05")
